connector/tests/common.py

```diff
@@ -1,47 +1,63 @@
+    load stub components::
 
-      'odoo.addons.magentoerpconnect.stock_picking.export_picking_done'
 
-    This is a context manager, all the calls made to the job function in
-    job_path inside the context manager will be executed synchronously.
+        from odoo.addons.connector_magento.models.product.importer import (
+            CatalogImageImporter
+        )
 
-    .. note:: It uses :meth:`mock.patch` so it has the same pitfall
-              regarding the python path.  If the mock seems to have no
-              effect, read `Where to patch
-              <http://www.voidspace.org.uk/python/mock/patch.html#where-to-patch>`_
-              in the mock documentation.
+        class TestImportProductImage(MagentoSyncTestCase):
+
+            def test_import_images(self):
+
+                # create a Stub component replacing the normal adapter
+                # (it has the same '_usage')
+                class StubProductAdapter(Component):
+
+                    _name = 'stub.product.adapter'
+                    _collection = 'magento.backend'
+                    _usage = 'backend.adapter'
+                    _apply_on = 'magento.product.product'
+
+                    def get_images(self, id, storeview_id=None):
+                        return [...]
+
+                # build the Stub and the component we want to test
+                self._build_components(StubProductAdapter,
+                                       CatalogImageImporter)
+
+                work = WorkContext(model_name='magento.product.product',
+                                   collection=self.backend,
+                                   components_registry=self.comp_registry)
+                image_importer = work.component_by_name(
+                    'magento.product.image.importer'
+                )
+                # the image importer will lookup the 'backend.adapter'
+                # component, which will our Stub, and call get_images()
+                image_importer.run(111)
+                # ... asserts below
 
     """
-    job_module, job_name = job_path.rsplit('.', 1)
-    module = importlib.import_module(job_module)
-    job_func = getattr(module, job_name, None)
-    assert job_func, "The function %s must exist in %s" % (job_name,
-                                                           job_module)
-
-    def clean_args_for_func(*args, **kwargs):
-        # remove the special args reserved to '.delay()'
-        kwargs.pop('priority', None)
-        kwargs.pop('eta', None)
-        kwargs.pop('model_name', None)
-        kwargs.pop('max_retries', None)
-        kwargs.pop('description', None)
-        job_func(*args, **kwargs)
-
-    with mock.patch(job_path) as patched_job:
-        # call the function directly instead of '.delay()'
-        patched_job.delay.side_effect = clean_args_for_func
-        yield patched_job
+
+    def setUp(self):
+        super(ConnectorTransactionCase, self).setUp()
```

Comment:
@lmignon but `_register_hook` register components in the global component registry for the database, here I want them in the test `comp_registry`. I extracted the content of `_register_hook` in a `build_registry` so we can use it from the setup: 834b351

Indeed, it works well and is more extensible!

Regarding `setUpClass`,  we need to reset the test component registry between each test, not for a whole class. The reason is that if inside a test you add a new component, it should no longer be in the registry for the next test.

```python
    def test_register(self):
        """ Able to register components in components registry """
        class Component1(Component):
            _name = 'component1'

        class Component2(Component):
            _name = 'component2'

        # build the 'final classes' for the components and check that we find
        # them in the components registry
        Component1._build_component(self.comp_registry)
        Component2._build_component(self.comp_registry)
        self.assertEquals(
            ['base', 'component1', 'component2'],
            list(self.comp_registry)
        )

    def test_inherit_bases(self):
        """ Check __bases__ of Component with _inherit """
        class Component1(Component):
            _name = 'component1'

        class Component2(Component):
            _inherit = 'component1'

        class Component3(Component):
            _inherit = 'component1'

        Component1._build_component(self.comp_registry)
        Component2._build_component(self.comp_registry)
        Component3._build_component(self.comp_registry)
        self.assertEquals(
            (Component3,
             Component2,
             Component1,
             self.comp_registry['base']),
            self.comp_registry['component1'].__bases__
        )
```